### PR TITLE
Kokkos sync: mainly add CUDA ASM with memory scope

### PR DESCRIPTION
**Pull request by crtrott**
This syncs changes which were made on the Kokkos side. Largely this is fixing corner cases with some compilers, some outright mistakes (like in HIP where a wrong argument is given before to an instantiation macro), and adding of `dont_deduce_this_parameter` for most of the `compare_exchange` stuff. It also updates the SYCL side of atomics, using updated capabilities in the DPC++ compiler.

### CODE REVIEW

**Comment by crtrott** [comment]
Note I will follow this up with a new version of the ScopeCaller PR. I also have an update on the Kokkos side which allows us to build Kokkos against an externally installed desul.

File: atomics/include/desul/atomics/Compare_Exchange_CUDA.hpp [hunk 1]
```diff
@@ -59,10 +59,11 @@ __device__ inline void atomic_thread_fence(MemoryOrderSeqCst, MemoryScopeNode) {
...
 // do NOT have a way of having the code included for clang only when the CC is smaller
-// than 700
-// But on Clang the device side symbol list must be independent of __CUDA_ARCH__
-#if (defined(__CUDA_ARCH__) && (__CUDA_ARCH__ < 700)) || \
-    (!defined(__NVCC__) && defined(DESUL_CUDA_ARCH_IS_PRE_VOLTA))
+// than 700 But on Clang the device side symbol list must be independent of
+// __CUDA_ARCH__
+// FIXME temporary fix for https://github.com/kokkos/kokkos/issues/4390
+#if defined(__CUDA_ARCH__) && (__CUDA_ARCH__ < 700) || \
+    (!defined(__NVCC__) && defined(DESUL_CUDA_ARCH_IS_PRE_VOLTA) && 0)
```
Comment (dalg24) [inline atomics/include/desul/atomics/Compare_Exchange_CUDA.hpp:66]: Undo these changes.

File: atomics/include/desul/atomics/SYCL.hpp [hunk 2]
```diff
@@ -17,17 +17,69 @@ SPDX-License-Identifier: (BSD-3-Clause)
 
 namespace desul {
 
+// FIXME_SYCL We need to either use generic_space or figure out how to check for the
+// correct adress space in a SYCL-portable way.
+#ifndef __NVPTX__
```
Comment (dalg24) [inline atomics/include/desul/atomics/SYCL.hpp:22]: Undo.  These are the changes from #48

File: atomics/include/desul/atomics/SYCLConversions.hpp [hunk 3]
```diff
@@ -80,20 +80,15 @@ struct DesulToSYCLMemoryScope<MemoryScopeSystem, extended_namespace> {
       sycl_memory_scope<extended_namespace>::system;
 };
 
-template <class T, class MemoryOrder, class MemoryScope>
```
Comment (dalg24) [inline atomics/include/desul/atomics/SYCLConversions.hpp:83]: Undo (already proposed in #48)

File: atomics/include/desul/atomics/HIP.hpp [hunk 4]
```diff
@@ -141,8 +141,8 @@ DESUL_IMPL_HIP_HOST_FALLBACK_ATOMIC_FUN_INTEGRAL(max, Max)
 DESUL_IMPL_HIP_HOST_FALLBACK_ATOMIC_FUN_FLOATING_POINT(add, Add)
 DESUL_IMPL_HIP_HOST_FALLBACK_ATOMIC_FUN_FLOATING_POINT(sub, Sub)
 
-DESUL_IMPL_HIP_HOST_FALLBACK_ATOMIC_FUN(inc_mod, IncModOper, unsigned int)
-DESUL_IMPL_HIP_HOST_FALLBACK_ATOMIC_FUN(dec_mod, DecModOper, unsigned int)
+DESUL_IMPL_HIP_HOST_FALLBACK_ATOMIC_FUN(inc_mod, IncMod, unsigned int)
+DESUL_IMPL_HIP_HOST_FALLBACK_ATOMIC_FUN(dec_mod, DecMod, unsigned int)
```
Comment (dalg24) [inline atomics/include/desul/atomics/HIP.hpp:145]: Ugggg

File: atomics/include/desul/atomics/Compare_Exchange_SYCL.hpp [hunk 5]
```diff
@@ -27,13 +27,131 @@ inline void atomic_thread_fence(MemoryOrder, MemoryScope) {
       Impl::DesulToSYCLMemoryScope<MemoryScope, /*extended namespace*/ false>::value);
 }
 
+// FIXME_SYCL We need to either use generic_space or figure out how to check for the
```
Comment (dalg24) [inline atomics/include/desul/atomics/Compare_Exchange_SYCL.hpp:30]: Undo (belongs to #48)

File: atomics/include/desul/atomics/Compare_Exchange_OpenMP.hpp [hunk 6]
```diff
@@ -7,6 +7,10 @@ SPDX-License-Identifier: (BSD-3-Clause)
 */
 #ifndef DESUL_ATOMICS_COMPARE_EXCHANGE_OPENMP_HPP_
 #define DESUL_ATOMICS_COMPARE_EXCHANGE_OPENMP_HPP_
+#include <omp.h>
+
+#include <cstdio>
```
Comment (dalg24) [inline atomics/include/desul/atomics/Compare_Exchange_OpenMP.hpp:12]: I don't see why we should include that header.

**Review by dalg24 — CHANGES_REQUESTED** [review]
Remove all the changes from #48
We need to discuss the `Impl::dont_deduce_this_parameter_t` changes.  I am not convinced we should do that.

File: atomics/include/desul/atomics/Compare_Exchange_CUDA.hpp (hunk 1 above)
Comment (crtrott) [reply atomics/include/desul/atomics/Compare_Exchange_CUDA.hpp:66]: not doing this change means Kokkos will not pass testing

File: atomics/include/desul/atomics/Compare_Exchange_OpenMP.hpp (hunk 6 above)
Comment (crtrott) [reply atomics/include/desul/atomics/Compare_Exchange_OpenMP.hpp]: removed it

**Comment by crtrott** [comment]
The alignment change is necessary since the gcc functions we use need alignment in addition to size matching. This is for example a problem for `complex<float>` where the compare_exchange would fail at runtime if called with a 4 byte alignment pointer.

**Comment by crtrott** [comment]
I added the predicate/isglobal stuff here

**Comment by crtrott** [comment]
@trws can you take a look at this?

**Comment by trws** [comment]
I think I'm happy here, @dalg24 is there anything remaining you're concerned about?

**Review by dalg24 — APPROVED** [review]
